File: backend/src/main/java/com/lifey/trainer/service/OccurrenceGenerator.java
package com.lifey.trainer.service;

import com.lifey.trainer.Recurrence;
import com.lifey.trainer.exception.EmptyRecurrenceException;
import com.lifey.trainer.exception.ScheduleHorizonExceededException;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.util.Arrays;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Pure date arithmetic for materializing a {@code WorkoutSchedule}'s occurrences
 * (docs/personal_trainer/09-utemezett-edzesek-domain-backend.md). No time zone
 * involved anywhere: everything is {@link LocalDate}.
 */
final class OccurrenceGenerator {

    /** Sanity cap — daily for ~3 months tops out around 92; 100 leaves headroom without being unbounded. */
    static final int MAX_OCCURRENCES = 100;

    private OccurrenceGenerator() {
    }

    /**
     * @throws EmptyRecurrenceException       if the recurrence produces no dates in range
     * @throws ScheduleHorizonExceededException if it produces more than {@link #MAX_OCCURRENCES}
     */
    static List<LocalDate> generate(Recurrence recurrence, List<DayOfWeek> daysOfWeek, LocalDate startDate, LocalDate endDate) {
        List<LocalDate> dates = switch (recurrence) {
            case ONCE -> List.of(startDate);
            case DAILY -> startDate.datesUntil(endDate.plusDays(1)).toList();
            case WEEKLY -> {
                Set<DayOfWeek> selected = EnumSet.copyOf(daysOfWeek);
                yield startDate.datesUntil(endDate.plusDays(1))
                        .filter(d -> selected.contains(d.getDayOfWeek()))
                        .toList();
            }
        };

        if (dates.isEmpty()) {
            throw new EmptyRecurrenceException("Recurrence produces no occurrences in the given date range");
        }
        if (dates.size() > MAX_OCCURRENCES) {
            throw new ScheduleHorizonExceededException("Recurrence would create " + dates.size()
                    + " occurrences, more than the " + MAX_OCCURRENCES + " sanity cap");
        }
        return dates;
    }

    /** {@code MONDAY} -> {@code "MON"}, comma-joined, e.g. {@code "MON,THU"}. */
    static String formatDaysOfWeek(List<DayOfWeek> daysOfWeek) {
        return daysOfWeek.stream().map(OccurrenceGenerator::toCode).collect(Collectors.joining(","));
    }

    /** Inverse of {@link #formatDaysOfWeek}; {@code null}/blank input yields an empty list. */
    static List<DayOfWeek> parseDaysOfWeek(String csv) {
        if (csv == null || csv.isBlank()) {
            return List.of();
        }
        return Arrays.stream(csv.split(","))
                .map(OccurrenceGenerator::fromCode)
                .toList();
    }

    private static String toCode(DayOfWeek day) {
        return day.name().substring(0, 3);
    }

    private static DayOfWeek fromCode(String code) {
        String trimmed = code.trim();
        return Arrays.stream(DayOfWeek.values())
                .filter(day -> toCode(day).equals(trimmed))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("Not a recognized day-of-week code: " + trimmed));
    }
}

File: mobile/lib/features/trainer/schedule/domain/occurrence_generator.dart
import 'schedule.dart';

/// Dart port of the backend's `OccurrenceGenerator`
/// (docs/personal_trainer/09-utemezett-edzesek-domain-backend.md).
///
/// The server materializes every occurrence the moment a schedule is created;
/// this exists so the form can say **before** submitting how many dates that
/// will be and where they end ("this creates 12 sessions, Aug 5 – Oct 28",
/// frame F4). It must therefore agree with the server exactly — same
/// inclusive bounds, same cap, same "empty is not allowed" rule.
///
/// No time zone anywhere: like the backend, this is pure date arithmetic on
/// local calendar days.

/// Sanity cap. Daily for ~3 months tops out near 92; the backend refuses
/// anything past this, so the form stops the trainer here rather than letting
/// the request fail.
const int maxScheduleOccurrences = 100;

/// Why a recurrence cannot be turned into occurrences.
enum OccurrenceProblem {
  /// The rule matches no day in the range — e.g. "every Monday" over a range
  /// with no Monday in it.
  none,

  /// More dates than the backend will accept.
  tooMany,
}

/// The dates a schedule would create, plus what is wrong if it would create
/// nothing usable.
class OccurrencePreview {
  const OccurrencePreview(this.dates, {this.problem});

  final List<DateTime> dates;
  final OccurrenceProblem? problem;

  bool get isValid => problem == null && dates.isNotEmpty;
  int get count => dates.length;
  DateTime? get first => dates.isEmpty ? null : dates.first;
  DateTime? get last => dates.isEmpty ? null : dates.last;
}

/// Every date the given rule produces, inclusive of both bounds.
///
/// [endDate] is ignored for [ScheduleRecurrence.once], exactly as the backend
/// ignores it: a one-off happens on its start date and nowhere else.
OccurrencePreview generateOccurrences({
  required ScheduleRecurrence recurrence,
  required List<ScheduleWeekday> daysOfWeek,
  required DateTime startDate,
  required DateTime endDate,
}) {
  final start = _dateOnly(startDate);
  final end = recurrence == ScheduleRecurrence.once ? start : _dateOnly(endDate);

  final dates = <DateTime>[];
  switch (recurrence) {
    case ScheduleRecurrence.once:
      dates.add(start);
    case ScheduleRecurrence.daily:
      for (var day = start; !day.isAfter(end); day = _nextDay(day)) {
        dates.add(day);
      }
    case ScheduleRecurrence.weekly:
      final selected = daysOfWeek.map((d) => d.isoNumber).toSet();
      for (var day = start; !day.isAfter(end); day = _nextDay(day)) {
        if (selected.contains(day.weekday)) dates.add(day);
      }
  }

  if (dates.isEmpty) {
    return const OccurrencePreview([], problem: OccurrenceProblem.none);
  }
  if (dates.length > maxScheduleOccurrences) {
    return OccurrencePreview(dates, problem: OccurrenceProblem.tooMany);
  }
  return OccurrencePreview(dates);
}

DateTime _dateOnly(DateTime date) => DateTime(date.year, date.month, date.day);

/// Steps a calendar day, not 24 hours: adding a `Duration` would land an hour
/// early or late across a daylight-saving boundary and silently drop or
/// duplicate a date.
DateTime _nextDay(DateTime date) =>
    DateTime(date.year, date.month, date.day + 1);
